File: src/core/data/repositories/aiChallengeHistoryRepository.ts
import { getDatabase } from '../database';
import type {
  AiChallengeCloseReason,
  AiChallengeHistoryEntry,
  AiChallengeMetric,
  AiChallengeTier,
  AiMissionSubTargets,
  NewAiChallengeHistoryEntry
} from '../../domain';
import { getCurrentProfileId } from './profileRepository';

export const aiChallengeHistoryRepository = {
  async create(entry: NewAiChallengeHistoryEntry): Promise<void> {
    const database = await getDatabase();
    const profileId = await getCurrentProfileId();
    if (entry.profileId !== profileId) {
      throw new Error('History entry profile_id does not match active profile.');
    }

    const subJson = entry.subTargets != null ? JSON.stringify(entry.subTargets) : null;
    const unitMin =
      entry.unitMinMinutes !== undefined && entry.unitMinMinutes !== null
        ? entry.unitMinMinutes
        : null;

    await database.execute(
      `INSERT INTO ai_challenge_history (
        profile_id, tier, title, description, metric, target, progress_at_close,
        close_reason, closed_at, original_created_at, original_expires_at,
        sub_targets_json, unit_min_minutes
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        entry.profileId,
        entry.tier,
        entry.title,
        entry.description,
        entry.metric,
        entry.target,
        entry.progressAtClose,
        entry.closeReason,
        entry.closedAt,
        entry.originalCreatedAt,
        entry.originalExpiresAt,
        subJson,
        unitMin
      ]
    );
  },

  async getRecent(tier: AiChallengeTier, limit = 50): Promise<AiChallengeHistoryEntry[]> {
    const database = await getDatabase();
    const profileId = await getCurrentProfileId();
    const rows = await database.select<any[]>(
      `SELECT * FROM ai_challenge_history
       WHERE profile_id = ? AND tier = ?
       ORDER BY closed_at DESC
       LIMIT ?`,
      [profileId, tier, limit]
    );
    return rows.map(mapHistoryRow);
  },

  async pruneOldestBeyondLimit(tier: AiChallengeTier, limit = 50): Promise<void> {
    const database = await getDatabase();
    const profileId = await getCurrentProfileId();
    await database.execute(
      `DELETE FROM ai_challenge_history
       WHERE id IN (
         SELECT id FROM (
           SELECT id,
                  ROW_NUMBER() OVER (
                    PARTITION BY profile_id, tier
                    ORDER BY closed_at DESC
                  ) AS rn
           FROM ai_challenge_history
           WHERE profile_id = ? AND tier = ?
         ) AS ranked
         WHERE rn > ?
       )`,
      [profileId, tier, limit]
    );
  }
};

function mapHistoryRow(row: any): AiChallengeHistoryEntry {
  const closeRaw = row.close_reason as string;
  const closeReason: AiChallengeCloseReason =
    closeRaw === 'replaced' || closeRaw === 'expired' || closeRaw === 'completed'
      ? closeRaw
      : 'completed';

  let subTargets: AiMissionSubTargets | null | undefined;
  const st = row.sub_targets_json;
  if (typeof st === 'string' && st.trim().length > 0) {
    try {
      const parsed = JSON.parse(st) as AiMissionSubTargets;
      if (parsed && typeof parsed.mode === 'string' && typeof parsed.count === 'number') {
        subTargets = parsed;
      }
    } catch {
      subTargets = null;
    }
  }

  return {
    id: Number(row.id),
    profileId: Number(row.profile_id),
    tier: row.tier as AiChallengeTier,
    title: row.title,
    description: row.description,
    metric: row.metric as AiChallengeMetric,
    target: Number(row.target),
    progressAtClose: Number(row.progress_at_close),
    closeReason,
    closedAt: row.closed_at,
    originalCreatedAt: row.original_created_at,
    originalExpiresAt: row.original_expires_at,
    subTargets: subTargets ?? null,
    unitMinMinutes: row.unit_min_minutes != null ? Number(row.unit_min_minutes) : null
  };
}

File: study_tracker_flutter/lib/data/repositories/ai_challenge_history_repository.dart
import 'package:drift/drift.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../db/database.dart' as db;
import '../db/database_provider.dart';
import '../../domain/domain.dart' as domain;

class AiChallengeHistoryRepository {
  final db.AppDatabase _db;

  AiChallengeHistoryRepository(this._db);

  Future<int> _activeProfileId() async {
    final row = await (_db.select(_db.appSettings)
          ..where((t) => t.key.equals('currentProfileId')))
        .getSingleOrNull();
    return int.tryParse(row?.value ?? '1') ?? 1;
  }

  Future<void> create(domain.NewAiChallengeHistoryEntry entry) async {
    final profileId = await _activeProfileId();
    if (entry.profileId != profileId) {
      throw StateError('History entry profileId does not match active profile.');
    }
    await _db.into(_db.aiChallengeHistory).insert(
          db.AiChallengeHistoryCompanion.insert(
            profileId: entry.profileId,
            tier: entry.tier.name,
            title: entry.title,
            description: entry.description,
            metric: entry.metric.name,
            target: entry.target,
            progressAtClose: entry.progressAtClose,
            closeReason: entry.closeReason.name,
            closedAt: entry.closedAt,
            originalCreatedAt: entry.originalCreatedAt,
            originalExpiresAt: entry.originalExpiresAt,
            subTargetsJson: Value(entry.subTargets?.toJsonString()),
            unitMinMinutes: Value(entry.unitMinMinutes),
          ),
        );
  }

  Future<List<domain.AiChallengeHistoryEntry>> getRecent(
    domain.AiChallengeTier tier, {
    int limit = 50,
  }) async {
    final profileId = await _activeProfileId();
    final rows = await (_db.select(_db.aiChallengeHistory)
          ..where((t) => t.profileId.equals(profileId) & t.tier.equals(tier.name))
          ..orderBy([(t) => OrderingTerm.desc(t.closedAt)])
          ..limit(limit))
        .get();
    return rows.map(_mapRow).toList();
  }

  Future<void> pruneOldestBeyondLimit(
    domain.AiChallengeTier tier, [
    int limit = 50,
  ]) async {
    final profileId = await _activeProfileId();
    await _db.customStatement(
      r'''
DELETE FROM ai_challenge_history
WHERE id IN (
  SELECT id FROM (
    SELECT id,
           ROW_NUMBER() OVER (
             PARTITION BY profile_id, tier
             ORDER BY closed_at DESC
           ) AS rn
    FROM ai_challenge_history
    WHERE profile_id = ? AND tier = ?
  ) AS ranked
  WHERE rn > ?
)
''',
      [profileId, tier.name, limit],
    );
  }

  domain.AiChallengeHistoryEntry _mapRow(db.AiChallengeHistoryRow row) {
    return domain.AiChallengeHistoryEntry(
      id: row.id,
      profileId: row.profileId,
      tier: domain.AiChallengeTier.fromDb(row.tier),
      title: row.title,
      description: row.description,
      metric: domain.AiChallengeMetric.fromDb(row.metric),
      target: row.target,
      progressAtClose: row.progressAtClose,
      closeReason: domain.AiChallengeCloseReason.fromDb(row.closeReason),
      closedAt: row.closedAt,
      originalCreatedAt: row.originalCreatedAt,
      originalExpiresAt: row.originalExpiresAt,
      subTargets: domain.AiMissionSubTargets.tryParse(row.subTargetsJson),
      unitMinMinutes: row.unitMinMinutes,
    );
  }
}

final aiChallengeHistoryRepositoryProvider = Provider<AiChallengeHistoryRepository>(
  (ref) => AiChallengeHistoryRepository(ref.watch(databaseProvider)),
);
